clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+        SCHEDULE_BY_INSTRUCTION_QUANTUM,
+        /**
+         * Uses a synthetic schedule with pre-emption of an input to be replaced
+         * by another input on a given output stream when it uses up its
+         * quantum which is measured using the user's notion of time.
+         * This notion of time must be supplied by the user by calling
+         * dynamorio::drmemtrace::scheduler_tmpl_t::stream_t::report_time()
+         * periodically.
+         * The quantum size is specified by the `quantum_duration` field of
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::scheduler_options_t.
+         */
+        SCHEDULE_BY_TIME_QUANTUM,
+    };
+
+    /** Flags controlling aspects of the scheduler beyond the core scheduling. */
+    enum scheduler_flags_t {
+        /** Default behavior. */
+        SCHEDULER_DEFAULTS = 0,
+        // TODO i#5843: Decide and then describe whether the physaddr is in every
+        // record or obtained via an API call.
+        /**
+         * Whether physical addresses should be provided in addition to virtual
+         * addresses.
+         */
+        SCHEDULER_PROVIDE_PHYSICAL_ADDRESSES = 0x1,
+        /**
+         * Specifies that speculation should supply just NOP instructions.
+         */
+        SCHEDULER_SPECULATE_NOPS = 0x2,
+        // TODO i#5843: Add more speculation flags for other strategies.
+    };
+
+    /**
+     * Collects the parameters specifying how the scheduler should behave, outside
+     * of the workload inputs and the output count.
+     */
+    struct scheduler_options_t {
+        /** Constructs a default set of options. */
+        scheduler_options_t()
+        {
+        }
+        /** Constructs a set of options with the given type and strategy. */
+        scheduler_options_t(stream_type_t stream_type, schedule_strategy_t strategy)
+            : how_split(stream_type)
+            , strategy(strategy)
+        {
+        }
+
+        /** Size of the struct for binary-compatible additions. */
+        size_t struct_size = sizeof(scheduler_options_t);
+        /** The base mode of the scheduler with respect to input streams. */
+        stream_type_t how_split = STREAM_BY_SYNTHETIC_CPU;
+        /** The scheduling strategy. */
+        schedule_strategy_t strategy = SCHEDULE_BY_INSTRUCTION_QUANTUM;
+        /** Optional flags affecting scheduler behavior. */
+        scheduler_flags_t flags = SCHEDULER_DEFAULTS;
+        /**
+         * The scheduling quantum duration for preemption.  The units are either
+         * instructions or user-controlled time, depending on whether
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULE_BY_INSTRUCTION_QUANTUM
+         * or
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::SCHEDULE_BY_TIME_QUANTUM
+         * is set.
+         */
+        uint64_t quantum_duration = 10 * 1000 * 1000;
+        /**
+         * Whether input stream inter-dependencies should be considered when preempting
+         * input streams.
```

Comment:
I know more documentation is planned, but it would be helpful to describe what kinds of inter-dependencies are considered.  If there are multiple types of dependencies possible, maybe this should be an enum rather than a bool.

Reply:
Made an enum.